src/cast_core/test/CrudActorYamlTestRunner.cpp

```diff
@@ -0,0 +1,269 @@
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <vector>
+
+#include <boost/algorithm/string/trim.hpp>
+#include <boost/exception/detail/exception_ptr.hpp>
+#include <boost/exception/diagnostic_information.hpp>
+
+
+#include <bsoncxx/builder/stream/document.hpp>
+#include <bsoncxx/json.hpp>
+#include <bsoncxx/types.hpp>
+
+#include <mongocxx/client.hpp>
+#include <mongocxx/stdx.hpp>
+#include <mongocxx/uri.hpp>
+
+#include <testlib/ActorHelper.hpp>
+#include <testlib/MongoTestFixture.hpp>
+#include <testlib/helpers.hpp>
+#include <testlib/yamlTest.hpp>
+#include <testlib/yamlToBson.hpp>
+
+namespace genny::testing {
+
+namespace {
+
+const char* DEFAULT_DB = "mydb";
+const char* DEFAULT_COLLECTION = "test";
+
+enum class RunMode { kNormal, kExpectedSetupException, kExpectedRuntimeException };
+
+RunMode convertRunMode(YAML::Node tcase) {
+    if (tcase["OutcomeData"] || tcase["OutcomeCounts"] || tcase["ExpectAllEvents"] ||
+        tcase["ExpectedCollectionsExist"]) {
+        return RunMode::kNormal;
+    }
+    if (tcase["Error"]) {
+        auto error = tcase["Error"].as<std::string>();
+        if (error == "InvalidSyntax") {
+            return RunMode::kExpectedSetupException;
+        }
+        return RunMode::kExpectedRuntimeException;
+    }
+    BOOST_THROW_EXCEPTION(std::logic_error("Invalid test-case"));
+}
+
+
+void requireExpectedCollectionsExist(mongocxx::pool::entry& client,
+                                     ApmEvents& events,
+                                     YAML::Node expectCollections) {
+    auto db = (*client)[DEFAULT_DB];
+    auto haystack = db.list_collection_names();
+
+    for (auto&& kvp : expectCollections) {
+        auto needle = kvp.first.as<std::string>();
+        auto expect = kvp.second.as<bool>();
+        auto actual = std::find(haystack.begin(), haystack.end(), needle) != haystack.end();
+        INFO("Expecting collection " << needle << "to exist? " << actual);
+        REQUIRE(expect == actual);
+    }
+}
+
+bool isNumeric(YAML::Node node) {
+    try {
+        node.as<long>();
+        return true;
+    } catch (const std::exception& x) {
+    }
+    return false;
+}
+
+// This isn't ideal. We want to assert only a subset of the keys in the ApmEvents
+// that we receive. Doing a subsetCompare(YAML::Node, bsoncxx::document::view) is nontrivial.
+// (Mostly fighting the compiler with yaml scalars versus the many numeric bson types.)
+// For now just assert a subset
+void requireEvent(ApmEvent& event, YAML::Node requirements) {
+    if (auto w = requirements["writeConcern"]["w"]; w) {
```

Comment:
I think ideally we'd remove the keys we've read and assert that there are no keys remaining in `requirements`. Not sure if usability matters that much though since this is all internal code.

Reply:
Ya I considered that but I didn't want to modify the input node (YAML::Node is basically a shared_ptr).